sys/posix/sockets/posix_sockets.c

```diff
@@ -0,0 +1,624 @@
+        case SOCK_STREAM:
+            if ((protocol == 0) || (protocol == IPPROTO_TCP)) {
+                return protocol;
+            }
+            else {
+                errno = EPROTOTYPE;
+            }
+            break;
+#endif
+#if defined(MODULE_CONN_UDP4) || defined(MODULE_CONN_UDP6)
+        case SOCK_DGRAM:
+            if ((protocol == 0) || (protocol == IPPROTO_UDP)) {
+                return protocol;
+            }
+            else {
+                errno = EPROTOTYPE;
+            }
+            break;
+#endif
+#if defined(MODULE_CONN_IP4) || defined(MODULE_CONN_IP6)
+        case SOCK_RAW:
+            return protocol;
+#endif
+        default:
+            break;
+    }
+    errno = EPROTONOSUPPORT;
+    return -1;
+}
+
+static int socket_close(int socket)
+{
+    socket_t *s;
+    int res = 0;
+    if ((unsigned)(socket - 1) > (SOCKET_POOL_SIZE - 1)) {
+        return -1;
+    }
+    mutex_lock(&_pool_mutex);
+    s = &_pool[socket];
+    s->domain = AF_INET;
+    switch (s->domain) {
+#if CONN_IPV4
+        case AF_INET:
+            switch (s->type) {
+#ifdef MODULE_CONN_UDP4
+                case SOCK_DGRAM:
+                    conn_udp4_close(&s->conn.udp4);
+                    break;
+#endif
+#ifdef MODULE_CONN_IP4
+                case SOCK_RAW:
+                    conn_ip4_close(&s->conn.raw4);
+                    break;
+#endif
+#ifdef MODULE_CONN_TCP4
+                case SOCK_STREAM:
+                    conn_tcp4_close(&s->conn.tcp4);
+                    break;
+#endif
+                default:
+                    res = -1;
+                    break;
+            }
+#endif
+#if CONN_IPV6
+        case AF_INET6:
+            switch (s->type) {
+#ifdef MODULE_CONN_UDP6
+                case SOCK_DGRAM:
+                    conn_udp6_close(&s->conn.udp6);
+                    break;
+#endif
+#ifdef MODULE_CONN_IP6
+                case SOCK_RAW:
+                    conn_ip6_close(&s->conn.raw6);
+                    break;
+#endif
+#ifdef MODULE_CONN_TCP6
+                case SOCK_STREAM:
+                    conn_tcp6_close(&s->conn.tcp6);
+                    break;
+#endif
+                default:
+                    res = -1;
+                    break;
+            }
+#endif
+        default:
+            res = -1;
+            break;
+    }
+    mutex_unlock(&_pool_mutex);
+    return res;
+}
+
+int socket(int domain, int type, int protocol)
+{
+    int res = 0;
+    socket_t *s;
+    mutex_lock(&_pool_mutex);
+    s = _get_free_socket();
+    if (s == NULL) {
+        errno = ENFILE;
+        mutex_unlock(&_pool_mutex);
+        return -1;
+    }
+    switch (domain) {
+#if CONN_IPV4
+        case AF_INET:
+#endif
+#if CONN_IPV6
+        case AF_INET6:
+#endif
+#if CONN_IPV4 || CONN_IPV6
+            s->domain = domain;
+            s->type = type;
+            if ((s->protocol = _choose_ipproto(type, protocol)) < 0) {
+                res = -1;
+            }
+            break;
+#endif
```

Comment:
From https://github.com/authmillenon/RIOT/commit/902d305123e6abb27ae4924ac8d5f37c75e5e6e2#commitcomment-12756195:

> how about:
>
>     case AF_INET:
>     case AF_INET6:
>         s->domain = domain;
>         s->type = type;
>         if ((s->protocol = _choose_ipproto(type, protocol)) < 0) {
>             res = -1;
>         }
>         break;

That's different from the current code in that this code segment will be entered even if the associated connection modules are not there. But I can put the `EAFNOSUPPORT` check for them to the top.